server/manager/src/main/java/org/apache/accumulo/manager/TabletGroupWatcher.java

```diff
@@ -924,6 +942,176 @@ private void mergeMetadataRecords(MergeInfo info) throws AccumuloException {
+            tabletExtent.prevEndRow(), deleteRange.prevEndRow());
+        ranges.add(new Range(tabletExtent.prevEndRow(), false, deleteRange.prevEndRow(), true));
+      }
+
+      // This covers the case of when a deletion range overlaps the last tablet. We need to create a
+      // range that excludes the deletion.
+      if (deleteRange.endRow() != null
+          && tabletMetadata.getExtent().contains(deleteRange.endRow())) {
+        Manager.log.trace("Fencing tablet {} files to ({},{}]", tabletExtent, deleteRange.endRow(),
+            tabletExtent.endRow());
+        ranges.add(new Range(deleteRange.endRow(), false, tabletExtent.endRow(), true));
+      }
+    } else {
+      Manager.log.trace(
+          "Fencing tablet {} files to itself because it does not overlap delete range",
+          tabletExtent);
+      ranges.add(tabletExtent.toDataRange());
+    }
+
+    return ranges;
+  }
+
+  private void updateMetadataRecordsForDelete(MergeInfo info) throws AccumuloException {
+    final KeyExtent range = info.getExtent();
+
+    String targetSystemTable = MetadataTable.NAME;
+    if (range.isMeta()) {
+      targetSystemTable = RootTable.NAME;
+    }
+
+    final AccumuloClient client = manager.getContext();
+
+    try (BatchWriter bw = client.createBatchWriter(targetSystemTable)) {
+      final Text startRow = range.prevEndRow();
+      final Text lastRow = range.endRow() != null
+          ? new Key(range.endRow()).followingKey(PartialKey.ROW).getRow() : null;
+
+      // Find the tablets that overlap the start and end of the deletion range
+      // Store in a Map so that if we have the same Tablet we only need to process once
+      final SortedMap<KeyExtent,TabletMetadata> tabletMetadatas = new TreeMap<>();
+      loadTabletMetadata(range.tableId(), startRow, ColumnType.PREV_ROW, ColumnType.FILES)
+          .ifPresent(tm -> tabletMetadatas.put(tm.getExtent(), tm));
+      loadTabletMetadata(range.tableId(), lastRow, ColumnType.PREV_ROW, ColumnType.FILES)
+          .ifPresent(tm -> tabletMetadatas.putIfAbsent(tm.getExtent(), tm));
+
+      for (TabletMetadata tabletMetadata : tabletMetadatas.values()) {
+        final KeyExtent keyExtent = tabletMetadata.getExtent();
+
+        // Check if this tablet needs to have its files fenced for the deletion
+        if (needsFencingForDeletion(info, keyExtent)) {
+          Manager.log.debug("Found overlapping keyExtent {} for delete, fencing files.", keyExtent);
+
+          // Create the ranges for fencing the files, this takes the place of
+          // chop compactions and splits
+          final List<Range> ranges = createRangesForDeletion(tabletMetadata, range);
+          Preconditions.checkState(!ranges.isEmpty(),
+              "No ranges found that overlap deletion range.");
+
+          // Go through and fence each of the files that are part of the tablet
+          for (Entry<StoredTabletFile,DataFileValue> entry : tabletMetadata.getFilesMap()
```

Comment:
So there are 2 parts to this, first whether it can happen and second if it is an issue.

In terms of can it happen, off the top of my head I'm not 100% sure but I would think this may be possible depending on previous merge or deletions with things being clipped. I think to avoid it we'd need to make sure to collapse ranges on merging/deletes when we do metadata updates. Right now the merge code updates files and copies existing files from merged tablets independently but doesn't try and collapse ranges or anything for duplicate files. It was easier to do the merging without trying to collapse ranges for now but could look at doing that. The fencing RFile iterator does collapse a list of ranges though.

In terms of if it's an issue, I don't think this scenario would cause any actual problems. If you end up with two files that are identical it would just do an update/add with an identical file and range but it would just replace so wouldn't end up with a duplicate or anything. Is that correct? If this is an issue then one option would be to move the tracking of the two sets of files (the newFiles and existingFile sets) earlier and create them before the for loop that iterates over the files for the tablet and just track all new/deleted files across entire tablet before doing mutations instead of resetting the sets for each file. My reason for doing it for each file was to keep memory usage lower and not have to track files for the entire tablet all at once.

Reply:
> If you end up with two files that are identical it would just do an update/add with an identical file and range but it would just replace so wouldn't end up with a duplicate or anything. Is that correct? 

That does seem correct, I can not think of any serious problems caused by this,  There could be information loss in the data file value, which is not a serious problem.

Thinking about this a bit more, I can not think of situation with merge where it would produce files with overlapping ranges.  When merge adds ranges to files, its based on tablet ranges which are disjoint.   When this new code adds 2 ranges to the same file in a tablet, those 2 ranges are disjoint.  So maybe its not something that would be expected to happen, still not sure though.

Reply:
Yeah, I can try and think if there's a scenario but if it does happen we don't lose the data at least and ranges are the same.